### DISCUSSION TITLE
Best way to fetch data with Next.js and SWR?

### DISCUSSION THREAD

**Discussion by hrbengtsen** [supabase/supabase#764] [Questions]
What is the best way to fetch data from supabase using SWR?

Currently I have gotten it to work by doing:
```js
const fetcher = (...args) => fetch(...args).then((res) => res.json());

const { user } = Auth.useUser();
const { data: recipes, error } = useSWR(
    user ? '/api/getRecipes?uid=' + user.id : null,
    fetcher,
    { refreshInterval: 1000 }
);
```

With api route:
```js
import { supabase } from '../../lib/initSupabase';

const getRecipes = async (req, res) => {
  const id = req.query.uid;

  const { data, error } = await supabase
    .from('recipes')
    .select('*')
    .eq('user_id', id)
    .order('id', true);

  if (error) return res.status(401).json({ error: error.message });
  return res.status(200).json(data);
};

export default getRecipes;
```

Is this a good approach? Or is there a better way to do it?

Also I don't know much about Row-level security, but was wondering if it's possible to use it with SWR? I've tried, but it doesn't work, which I'm guessing is because the api routes are server-side.

**Comment by Team-SEO** [comment]
heres what ive got:

fetcher
`import axios from "axios";

const fetcher = (url) => axios.get(url).then((res) => res.data);

export default fetcher;`

SWR
`import useSWR from "swr";
import fetcher from "src/lib/utils/swr/fetcher";

const useUser= () => {
  const { data, error, mutate } = useSWR("/api/auth/session/user", fetcher);

  return {
    user: data,
    isUserErr: error,
    isUserLoading: !error && !data,
    mutateUser: mutate,
  };
};

export default useUser;`

API
`const { data, error } = await supabase
        .from("profiles")
        .select("*")
        .match({ uuid: userId });`

I left out some details as it is not relevant to this, but you see the useSWR hook calls the api with the fetcher, the api calls the database and it is brought to your front end where ever you use the useSWR hook

sorry the code blocks didnt all work

**Comment by psteinroe** [comment]
If you plan to use swr with supabase, be sure to check out [Supabase Cache Helpers](https://github.com/psteinroe/supabase-cache-helpers). It integrates supabase-js with swr and comes with mutation, subscription and convenience hooks for pagination as well as storage support. All logic is split up in reusable packages, hence adding support for `react-query` should be quite straight forward too.

```tsx
   // Define the query. Check out the other query hooks for pagination etc.
    const { data, count } = useQuery(
    client
        .from("contact")
        .select("id,username,ticket_number", { count: "exact" })
        .eq("username", "myusername"),
    "multiple",
    {
        revalidateOnFocus: false,
        revalidateOnReconnect: false,
    }
    );

    // Define the mutation. Update, Upsert, and Delete are also supported.
    const [insert] = useInsertMutation(client.from("contact"), "single");

    // Subscriptions are also supported.
    const { status } = useSubscription(
        client.channel('random'),
        {
            event: "*",
            table: "contact",
            schema: "public",
        },
        ["id"],
        { callback: (payload) => console.log(payload) }
    );

```
